modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.client/src/main/scala/edu/gemini/seqexec/web/client/model/SeqexecCircuit.scala

```diff
@@ -180,14 +201,85 @@ class GlobalLogHandler[M](modelRW: ModelRW[M, GlobalLog]) extends ActionHandler(
 }
 
 /**
-  * Handles actions related to the changing the selection of the displayed sequence
+  * Handles the WebSocket connection and performs reconnection if needed
+  */
+class WebSocketHandler[M](modelRW: ModelRW[M, Option[WebSocket]]) extends ActionHandler(modelRW) {
+  implicit val runner = new RunAfterJS
+  val logger = Logger.getLogger(this.getClass.getSimpleName)
+  // Reconfigure to avoid sending ajax events in this logger
+  logger.setUseParentHandlers(false)
+  logger.addHandler(new ConsoleHandler(Level.FINE))
+
+  // Makes a websocket connection and setups event listeners
+  def webSocket(nextDelay: Int) = Future[Action] {
+    import org.scalajs.dom.document
+
+    val host = document.location.host
+    val url = s"ws://$host/api/seqexec/events"
+
+    def onOpen(e: Event): Unit = {
+      logger.info(s"Connected to $url")
+      SeqexecCircuit.dispatch(Connected)
+    }
+
+    def onMessage(e: MessageEvent): Unit = {
+      \/.fromTryCatchNonFatal(read[SeqexecEvent](e.data.toString)) match {
+        case \/-(event) => SeqexecCircuit.dispatch(NewSeqexecEvent(event))
+        case -\/(t)     => println(s"Error decoding event ${t.getMessage}")
+      }
+    }
+
+    def onError(e: ErrorEvent): Unit = {
+      // Unfortunately reading the event is not cross-browser safe
+      logger.severe("Error on websocket")
+    }
+
+    def onClose(e: CloseEvent): Unit =
+      // Increase the delay to get exponential backoff with a minimum of 200ms and a max of 1m
+      SeqexecCircuit.dispatch(ConnectionClosed(math.min(60000, math.max(200, nextDelay * 2))))
+
+    val ws = new WebSocket(url)
+    ws.onopen = onOpen _
+    ws.onmessage = onMessage _
+    ws.onerror = onError _
+    ws.onclose = onClose _
+    Connecting(ws)
+  }.recover {
+    case e: Throwable => NoAction
+  }
```

Comment:
This looks like perfectly respectable callback style code, the typical case to deal with websockets in JS. This is not large enough to become a significant issue, but I was wondering if it'd be too much trouble, or possible at all, to use a more monadic API, like with `Task.async`. It also would set the tone for a more monadic style in future commits minimizing the risk of ending up in callback hell.